### PR TITLE
Allow to run migrations on application start up

### PR DESCRIPTION
**Pull request by weiznich**
* Allow to run migrations on application start up without depending on migrations directory
* Add functionality to generate the needed code from the current migrations directory

### CODE REVIEW

File: diesel/src/migrations/migration.rs [hunk 1]
```diff
@@ -1,18 +1,13 @@
...
 use std::path::{Path, PathBuf};
 
-pub trait Migration {
-    fn version(&self) -> &str;
-    fn run(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
-    fn revert(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
-}
 
-pub fn migration_from(path: PathBuf) -> Result<Box<Migration>, MigrationError> {
+pub fn migration_from(path: PathBuf) -> Result<SqlFileMigration, MigrationError> {
```
Comment (sgrif) [inline diesel/src/migrations/migration.rs]: The signature of this shouldn't change. We keep these boxed for a reason.

File: diesel/src/migrations/migration.rs [hunk 2]
```diff
@@ -49,7 +44,7 @@ fn version_from_path(path: &Path) -> Result<String, MigrationError> {
 use std::fs::File;
 use std::io::Read;
 
-struct SqlFileMigration(PathBuf, String);
+pub struct SqlFileMigration(PathBuf, String);
```
Comment (sgrif) [inline diesel/src/migrations/migration.rs]: This shouldn't need to change.

File: diesel/src/migrations/migration.rs [hunk 3]
```diff
@@ -65,10 +60,27 @@ impl Migration for SqlFileMigration {
...
 }
 
-fn run_sql_from_file(conn: &SimpleConnection, path: &Path) -> Result<(), RunMigrationsError> {
+
+pub fn programmatic_migration_from(path: PathBuf) -> Result<String, MigrationError> {
+    if valid_sql_migration_directory(&path) {
+        let version = try!(version_from_path(&path));
+        let up = try!(sql_from_file(&path.join("up.sql")));
+        let down = try!(sql_from_file(&path.join("down.sql")));
+        Ok(format!("ProgrammaticMigration{{ version:\"{}\", up:\"{}\", down: \"{}\"}}",version.trim(),up.trim(),down.trim()))
```
Comment (sgrif) [inline diesel/src/migrations/migration.rs]: Code shouldn't get constructed by string munging. Can you move this to an annotation in codegen?

File: diesel/src/migrations/mod.rs [hunk 4]
```diff
@@ -91,27 +93,128 @@ use std::path::{PathBuf, Path};
...
+/// include!(concat!(env!("OUT_DIR"), "/migrations.rs"));
+/// 
+/// use self::migrations::run;
+/// let connection = establish_connection();
+/// run(&conn).unwrap();
+/// ```
+/// See the [module level documentation](index.html) for information on how migrations should be
+/// structured, and where Diesel will look for them by default.
+
+pub fn generate_migrations_module(dst: &Path) -> Result<(), RunMigrationsError> {
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Again, code generation doesn't belong in the main Diesel crate, nor should we be writing to files. This should happen inside of codegen.

File: diesel/src/migrations/mod.rs [hunk 5]
```diff
@@ -91,27 +93,128 @@ use std::path::{PathBuf, Path};
...
+    let mut f = try!(File::create(dst));
+    try!(f.write_all(&migration_module.as_bytes()));
+    Ok(())
+}
+
+
 #[doc(hidden)]
-pub fn run_pending_migrations_in_directory<Conn>(conn: &Conn, migrations_dir: &Path, output: &mut Write)
-    -> Result<(), RunMigrationsError> where
+pub fn run_pending_migrations_form_iter<Conn, T, M>(conn: &Conn, all_migrations:T, output: &mut Write)-> Result<Vec<String>, RunMigrationsError> where
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: This method has a typo.

Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Please make sure the code adheres to the [style guidelines](https://github.com/sgrif/diesel/issues/120#issuecomment-174222639)

File: diesel/src/migrations/mod.rs [hunk 6]
```diff
@@ -91,27 +93,128 @@ use std::path::{PathBuf, Path};
...
+    run_pending_migrations_form_iter(conn,all_migrations.into_iter(),output)
+}
+
+fn build_mirgations_module() -> Result<String, MigrationError>{
+
+    let migrations_dir = try!(find_migrations_directory());
+    build_migrations_module_form_directory(&migrations_dir)
+}
+
+fn join<S:ToString> (l: Vec<S>, sep: &str) -> String{
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Is there a reason we can't use join from the standard library?

File: diesel/src/migrations/mod.rs [hunk 7]
```diff
@@ -91,27 +93,128 @@ use std::path::{PathBuf, Path};
...
+                  |a,b| if a.len()>0 {a+sep} else {a} + &b.to_string()
+                  )
+}
+
+fn build_migrations_module_form_directory(migrations_dir: &Path) -> Result<String, MigrationError>
+{
+    use self::migration::programmatic_migration_from;
+
+    let migrations = try!(migrations_in_directory(migrations_dir,&programmatic_migration_from));
+    Ok(format!("pub mod migrations {{
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Same thing here. We shouldn't be creating code using string munging, this needs to go in codegen.

File: diesel/src/migrations/mod.rs [hunk 8]
```diff
@@ -238,6 +346,39 @@ fn search_for_migrations_directory(path: &Path) -> Result<PathBuf, MigrationErro
...
 
+#[doc(hidden)]
+pub trait Migration {
+    fn version(&self) -> &str;
+    fn run(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
+    fn revert(&self, conn: &SimpleConnection) -> Result<(), RunMigrationsError>;
+}
+
+#[doc(hidden)]
+pub struct ProgrammaticMigration {
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: There's no reason that this needs to be in the main crate, we can stick this in the module generated for the users.

File: diesel_compile_tests/tests/compile_tests.rs [hunk 9]
```diff
@@ -23,4 +23,5 @@ fn run_mode(mode: &'static str) {
 #[test]
 fn compile_test() {
     run_mode("compile-fail");
+    run_mode("run-pass");
```
Comment (sgrif) [inline diesel_compile_tests/tests/compile_tests.rs]: Is there a reason these tests can't go in the main integration test suite?

**Comment by sgrif** [comment]
Also worth noting that all of the documentation examples you've added fails to run. That will need to be addressed as well before this can be merged. We have a script to help you run all of the test suites locally at `bin/test`, usually good to double check it before pushing.

**Comment by weiznich** [comment]
Thanks for the comments. I will rework this to use codegen instead.

**Comment by weiznich** [comment]
I've updated the pull request according to your comments.

**Comment by sgrif** [comment]
Sorry for the delay in responding to this. I've been on hiatus for a few weeks. I'll take another look at this soon -- just wanted to let you know I haven't forgotten

**Comment by weiznich** [comment]
Is there anything preventing this from being merged, except of time?

File: diesel_codegen/Cargo.toml [hunk 10]
```diff
@@ -17,7 +17,7 @@ syntex_syntax = { version = "^0.28.0", optional = true }
 [dependencies]
 syntex = { version = "^0.28.0", optional = true }
 syntex_syntax = { version = "^0.28.0", optional = true }
-diesel = { version = "0.5.0", default-features = false }
+diesel = { path="../diesel/", default-features = false }
```
Comment (sgrif) [inline diesel_codegen/Cargo.toml]: This shouldn't be required, it's overloaded in `.cargo/config`

File: diesel_tests/build.rs [hunk 11]
```diff
@@ -39,7 +39,7 @@ mod outer {
         let database_url = ::std::env::var("DATABASE_URL")
             .expect("DATABASE_URL must be set to run tests");
         let connection = PgConnection::establish(&database_url).unwrap();
-        let migrations_dir = migrations::find_migrations_directory().unwrap().join("postgresql");
+        let migrations_dir = migrations::find_migrations_directory().unwrap().join("../../migrations/postgresql");
```
Comment (sgrif) [inline diesel_tests/build.rs]: Why did this change?

File: diesel_tests/tests/migrations.rs [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+use diesel::{select, LoadDsl};
+use diesel::result::QueryResult;
+use diesel::connection::Connection;
+use diesel::expression::dsl::sql;
+use diesel::migrations::revert_migration;
+use schema::*;
+
+
+// FIXME: Remove the duplicates of this function once expression level attributes
+// are stable (I believe this is in 1.7)
```
Comment (sgrif) [inline diesel_tests/tests/migrations.rs]: Sorry this has taken so long to review, but 1.7 is stable now.

**Comment by sgrif** [comment]
There's a lot of style nitpicks as well, but I can just fix those manually. Will take another look today or tomorrow when I can go a bit more in depth.

File: diesel_tests/build.rs (hunk 11 above)
Comment (weiznich) [inline diesel_tests/build.rs]: Without this change the dummy migrations from diesel_test are used(They are added in this commit)

Comment (sgrif) [inline diesel_tests/build.rs]: Ah, makes sense. Honestly, with this feature added, I'm thinking we might just be able to cut most of the existing "runtime" migration functionality, and move it into `diesel_cli`. Packaging the migrations in the binary feels more rust-like. That doesn't need to be part of this PR though, this is fine for now.

Comment (sgrif) [inline diesel_tests/build.rs]: Actually, can you move the migrations that you added in this PR to a directory name other than `migrations` so this doesn't need to change?

File: diesel_tests/tests/migrations.rs (hunk 12 above)
Comment (mcasper) [inline diesel_tests/tests/migrations.rs]: Unfortunately expression level attributes weren't stabilized in 1.7 :/ 
https://github.com/rust-lang/rust/blob/1.7.0/src/libsyntax/feature_gate.rs#L232

Comment (sgrif) [inline diesel_tests/tests/migrations.rs]: For match arms it was.

Comment (weiznich) [inline diesel_tests/tests/migrations.rs]: We don't even need this here. Tests are only build for postgres or sqlite, not for both at the same time.

**Comment by weiznich** [comment]
Hopefully all style nitpicks are fixed. If not please leave a comment.

**Comment by sgrif** [comment]
Heading out of town, but I'll take a look when I have a chance.

File: diesel/src/migrations/mod.rs [hunk 13]
```diff
@@ -91,35 +92,47 @@ use std::path::{PathBuf, Path};
...
-pub fn run_pending_migrations_in_directory<Conn>(conn: &Conn, migrations_dir: &Path, output: &mut Write)
-    -> Result<(), RunMigrationsError> where
-        Conn: MigrationConnection,
-{
+pub fn run_pending_migrations_from_iter<Conn, T>(conn: &Conn,
+                                                 all_migrations: T,
+                                                 output: &mut Write)
+    -> Result<Vec<String>, RunMigrationsError>
+    where Conn: MigrationConnection,
+          T: Iterator<Item = Box<Migration>>, {
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Format for this should be:

```rust

pub fn run_pending_migrations_from_iter<Conn, T>(
    conn: &Conn,
    all_migrations: T,
    output: &mut Write,
) -> Result<Vec<String>, RunMigrationsError> where
    Conn: MigrationConnection,
    T: Iterator<Item = Box<Migration>>,
{
```

Comment (weiznich) [inline diesel/src/migrations/mod.rs]: Fixed

File: diesel/src/migrations/migration.rs [hunk 14]
```diff
@@ -1,13 +1,9 @@
 use connection::SimpleConnection;
-use super::{MigrationError, RunMigrationsError};
+use super::{Migration, MigrationError, RunMigrationsError, valid_sql_migration_directory,
+            version_from_path};
 
 use std::path::{Path, PathBuf};
 
-pub trait Migration {
```
Comment (sgrif) [inline diesel/src/migrations/migration.rs]: Why move this?

File: diesel/src/migrations/mod.rs [hunk 15]
```diff
@@ -91,66 +92,89 @@ use std::path::{PathBuf, Path};
...
+    Conn: MigrationConnection,
 {
     migration_with_version(ver)
         .map_err(|e| e.into())
-        .and_then(|m| revert_migration(conn, m, output))
+        .and_then(|ref m| revert_migration(conn, m, output))
 }
 
 #[doc(hidden)]
-pub fn run_migration_with_version<Conn>(conn: &Conn, ver: &str, output: &mut Write)
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Please don't reformat existing code that isn't being changed.

**Comment by sgrif** [comment]
It's unclear to me why any code in the migrations module should need to be run. It seems like this change needs to:

- Generate a struct with the up and down string, along with the version
- Implement the `Migration` trait
- Generate a constant with all the instances of this struct
- Generate a single function to which runs from that constant

File: diesel/src/migrations/migration.rs (hunk 14 above)
Comment (weiznich) [inline diesel/src/migrations/migration.rs]: this module is not public, so I can't use this trait in codegen

**Comment by weiznich** [comment]
Most changes in the migration module are needed to reuse the code in codegen (migration::migrations is not public). Furthermore I add a function to run migrations from a Iterator instead of running them from a directory.
The codegen part does more or less what you describe.

**Comment by sgrif** [comment]
Can we do some pub exports or make the module public instead of moving so
much code around?

**Comment by weiznich** [comment]
I revisited and minimized the changes in the migration module.

**Comment by sgrif** [comment]
Thank you. I will review/merge soon.

**Comment by sgrif** [comment]
Thank you for working on this. I've gone ahead with a separate implementation, though this PR was used as a reference. I'll make sure you're credited for the feature in the next release announcement.

**Pull request closed without merging**